kotlinx-coroutines-core/jvm/src/CoroutineContext.kt

```diff
@@ -213,19 +227,25 @@ internal actual class UndispatchedCoroutine<in T>actual constructor (
     }
 
     fun saveThreadContext(context: CoroutineContext, oldValue: Any?) {
+        threadLocalIsSet = true // Specify that thread-local is touched at all
         threadStateToRecover.set(context to oldValue)
     }
 
     fun clearThreadContext(): Boolean {
-        if (threadStateToRecover.get() == null) return false
-        threadStateToRecover.set(null)
+        if (threadLocalIsSet && threadStateToRecover.get() == null) {
+            threadStateToRecover.remove()
```

Comment:
I think such brittle changes should be heavily documented so that they don't get reverted by seemingly innocent edits. I *guess* the reason to call `threadStateToRecover.remove()` here is to completely eliminate the thread-local variable so that it doesn't get traversed (is this correct at all?), but there's no chance I'd understand the danger of removing this line if I didn't know or forgot about this issue.

Reply:
I've added some, could you please check if that is enough along with commit messages?